card/mod/01_core/set/all/name_validations.rb

```diff
@@ -0,0 +1,56 @@
+event :permit_codename, :validate,
+      on: :update, changed: :codename do
+  errors.add :codename, 'only admins can set codename' unless Auth.always_ok?
+  validate_uniqueness_of_codename
+end
+
+event :validate_uniqueness_of_codename do
+  return unless codename.present? && errors.empty? &&
+                Card.find_by_codename(codename).present?
+  errors.add :codename, "codename #{codename} already in use"
+end
+
+event :validate_name, :validate,
+      on: :save, changed: :name do
+  validate_legality_of_name
+  return if errors.any?
+  validate_uniqueness_of_name
+end
+
+event :validate_uniqueness_of_name do
+  # validate uniqueness of name
+  condition_sql = 'cards.key = ? and trash=?'
+  condition_params = [cardname.key, false]
+  unless new_record?
+    condition_sql << ' AND cards.id <> ?'
+    condition_params << id
+  end
+  if (c = Card.find_by(condition_sql, *condition_params))
+    errors.add :name, "must be unique; '#{c.name}' already exists."
+  end
```

Comment:
just noticing that this has similarities with fetch_from_db.  might be able to reuse some code...